### PR TITLE
ec2_instance exception handling and client cleanup

### PR DESCRIPTION
**Pull request by jillr**
##### SUMMARY
Catch botocore and client errors on all API calls
Pass boto client to functions, rather than creating new clients throughout the code

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
ec2_instance
ec2_instance_info

### CODE REVIEW

**Comment by jillr** [comment]
recheck

**Comment by tremble** [comment]
recheck

File: plugins/modules/ec2_instance.py [hunk 1]
```diff
@@ -1389,28 +1392,37 @@ def find_instances(ec2, ids=None, filters=None):
                 filters[key.replace("_", "-")] = filters.pop(key)
         params = dict(Filters=ansible_dict_to_boto3_filter_list(filters))
 
-    results = paginator.paginate(**params).search('Reservations[].Instances[]')
+    try:
+        results = paginator.paginate(**params).search('Reservations[].Instances[]')
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+        module.fail_json_aws(e, msg="Could not describe instances")
```
Comment (tremble) [inline plugins/modules/ec2_instance.py:1398]: Catching the Error here prevents it falling through the the jittered backoff and being retried it's probably worth  rejigging this a little:

```suggestion
    try:
        results = _describe_instances(ec2, **params)
    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
        module.fail_json_aws(e, msg="Could not describe instances")


@AWSRetry.jittered_backoff()
def _describe_instances(ec2, **params):   
    paginator = ec2.get_paginator('describe_instances')
    paginator.paginate(**params).search('Reservations[].Instances[]')
```

**Review by tremble — CHANGES_REQUESTED** [review]
Mostly looks good.

File: plugins/modules/ec2_instance.py [hunk 2]
```diff
@@ -1389,28 +1392,43 @@ def find_instances(ec2, ids=None, filters=None):
...
+    try:
+        results = _describe_instances(ec2, **params)
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+        module.fail_json_aws(e, msg="Could not describe instances")
+
+
+@AWSRetry.jittered_backoff()
+def _describe_instances(ec2, **params):   
+    paginator = ec2.get_paginator('describe_instances')
+    paginator.paginate(**params).search('Reservations[].Instances[]')
```
Comment (tremble) [inline plugins/modules/ec2_instance.py:1404]:
```suggestion
    results = paginator.paginate(**params).search('Reservations[].Instances[]')
```

File: plugins/modules/ec2_instance.py [hunk 3]
```diff
@@ -1389,28 +1392,43 @@ def find_instances(ec2, ids=None, filters=None):
...
+        results = _describe_instances(ec2, **params)
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+        module.fail_json_aws(e, msg="Could not describe instances")
+
+
+@AWSRetry.jittered_backoff()
+def _describe_instances(ec2, **params):   
+    paginator = ec2.get_paginator('describe_instances')
     results = paginator.paginate(**params).search('Reservations[].Instances[]')
     return list(results)
```
Comment (tremble) [inline plugins/modules/ec2_instance.py:1405]:
```suggestion
    try:
        results = _describe_instances(ec2, **params)
    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
        module.fail_json_aws(e, msg="Could not describe instances")
    return list(results)

@AWSRetry.jittered_backoff()
def _describe_instances(ec2, **params):   
    paginator = ec2.get_paginator('describe_instances')
    return paginator.paginate(**params).search('Reservations[].Instances[]')
```

File: plugins/modules/ec2_instance.py [hunk 4]
```diff
@@ -1389,28 +1392,42 @@ def find_instances(ec2, ids=None, filters=None):
...
-    results = paginator.paginate(**params).search('Reservations[].Instances[]')
+    try:
+        results = _describe_instances(ec2, **params)
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+        module.fail_json_aws(e, msg="Could not describe instances")
     return list(results)
 
+@AWSRetry.jittered_backoff()
+def _describe_instances(ec2, **params):   
+    paginator = ec2.get_paginator('describe_instances')
```
Comment (tremble) [inline plugins/modules/ec2_instance.py:1403]:
```suggestion
    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
        module.fail_json_aws(e, msg="Could not describe instances")
    return list(results)


@AWSRetry.jittered_backoff()
def _describe_instances(ec2, **params):
    paginator = ec2.get_paginator('describe_instances')
```

**Comment by tremble** [comment]
Based on the test failures it looks like we still have a race condition on creation.  Once the tests here pass I'll submit a separate PR to cover it.

**Review by tremble — APPROVED** [review]

**Pull request merged**